lenskit-slim/src/main/java/org/lenskit/slim/SLIMBuildContextProvider.java

```diff
@@ -0,0 +1,301 @@
+import org.lenskit.util.io.ObjectStream;
+import org.lenskit.util.math.Vectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * SLIM model build context provider
+ * Build the necessary context for SLIM model
+ * including user-item rating map, item neighbors map, item-item inner-products map and users' rated items
+ *
+ * @author <a href="http://www.grouplens.org">GroupLens Research</a>
+ */
+public class SLIMBuildContextProvider implements Provider<SLIMBuildContext> {
+    private static final Logger logger = LoggerFactory.getLogger(SLIMBuildContextProvider.class);
+
+
+    private final DataAccessObject dao;
+    private final ItemVectorNormalizer normalizer;
+    private final VectorSimilarity itemSimilarity;
+    private final Threshold threshold;
+    private final int minCommonUsers;
+    private final int modelSize;
+
+    @Inject
+    public SLIMBuildContextProvider(@Transient DataAccessObject dao,
+                                    @Transient ItemVectorNormalizer normlzr,
+                                    @Transient VectorSimilarity sim,
+                                    @Transient Threshold threshld,
+                                    @MinCommonUsers int minCI,
+                                    @SLIMModelSize int size) {
+        this.dao = dao;
+        normalizer = normlzr;
+        itemSimilarity = sim;
+        threshold = threshld;
+        minCommonUsers = minCI;
+        modelSize = size;
+
+    }
+
+
+    /**
+     * Construct the data objects needed by building slim model.
+     * @return The slimBuildContext.
+     */
+    @Override
+    public SLIMBuildContext get() {
+        Long2ObjectMap<Long2DoubleSortedMap> itemVectors = new Long2ObjectOpenHashMap<>();
+        Long2ObjectMap<LongSortedSet> userItems = new Long2ObjectOpenHashMap<>();
+        buildItemRatings(itemVectors, userItems);
+
+        Long2ObjectMap<Long2DoubleSortedMap> innerProducts = new Long2ObjectOpenHashMap<>();
+        buildItemItemInnerProducts(itemVectors, innerProducts);
+
+        Long2ObjectMap<LongSortedSet> itemNeighbors = new Long2ObjectOpenHashMap<>();
+        buildItemNeighbors(itemVectors, itemNeighbors);
+
+        return new SLIMBuildContext(itemVectors, itemNeighbors, innerProducts, userItems);
+    }
+
+
+    /**
+     * Build item rating matrix
+     * @param itemVectors mapping from item ids to (userId: rating) maps to be filled
+     * @param userItems mapping from user ids to sets of user rated items (to be filled)
+     */
+    private void buildItemRatings(Long2ObjectMap<Long2DoubleSortedMap> itemVectors,
+                                  Long2ObjectMap<LongSortedSet> userItems) {
+        Long2ObjectMap<LongSet> userItemsTemp = new Long2ObjectOpenHashMap<>();
+        try (ObjectStream<IdBox<List<Rating>>> stream = dao.query(Rating.class)

```

Comment:
We really need to use user-vector normalization for consistency. Also, we should use the `RatingVectorPDAO` instead of a data access object.

That means we will be processing ratings user-by-user instead of item-by-item. Once the final context is built, there will be no difference, but the loop here will need to change.